src/drivers/imu/adis16477/ADIS16477.cpp

```diff
@@ -255,79 +293,71 @@ ADIS16477::probe()
 {
 	DEVICE_DEBUG("probe");
 
-	reset();
-
 	// read product id (5 attempts)
 	for (int i = 0; i < 5; i++) {
+
+		if (reset() != PX4_OK) {
+			continue;
+		}
+
 		_product = read_reg16(PROD_ID);
 
 		if (_product == PROD_ID_ADIS16477) {
-			DEVICE_DEBUG("PRODUCT: %X", _product);
+			DEVICE_DEBUG("PRODUCT: %#X", _product);
 
-			if (self_test()) {
+			if (self_test_memory() && self_test_sensor()) {
 				return PX4_OK;
 
 			} else {
 				PX4_ERR("probe attempt %d: self test failed, resetting", i);
-				reset();
 			}
 
 		} else {
 			PX4_ERR("probe attempt %d: read product id failed, resetting", i);
-			reset();
 		}
 	}
 
 	return -EIO;
 }
 
-/* set sample rate for both accel and gyro */
-void
-ADIS16477::_set_sample_rate(uint16_t desired_sample_rate_hz)
-{
-
-}
-
-/* set the DLPF FIR filter tap. This affects both accelerometer and gyroscope. */
-void
-ADIS16477::_set_dlpf_filter(uint16_t desired_filter_tap)
+bool
+ADIS16477::self_test_memory()
 {
-	//modify_reg16(ADIS16477_SENS_AVG, 0x0007, desired_filter_tap);
+	DEVICE_DEBUG("self test memory");
 
-	/* Verify data write on the IMU */
+	// self test (global command bit 4)
+	uint8_t value[2] = {};
+	value[0] = (1 << 4);
+	write_reg16(GLOB_CMD, (uint16_t)value[0]);
+	up_mdelay(32); // Flash Memory Test Time
 
-	//if ((read_reg16(ADIS16477_SENS_AVG) & 0x0007) != desired_filter_tap) {
-	//	DEVICE_DEBUG("failed to set IMU filter");
-	//}
-}
+	// read DIAG_STAT to check result
+	uint16_t diag_stat = read_reg16(DIAG_STAT);
 
-/* set IMU to factory defaults. */
-void
-ADIS16477::_set_factory_default()
-{
-	//write_reg16(ADIS16477_GLOB_CMD, 0x02);
-}
+	if (diag_stat != 0) {
+		PX4_ERR("DIAG_STAT: %#X", diag_stat);
+		return false;
+	}
 
-/* set the gyroscope dynamic range */
-void
-ADIS16477::_set_gyro_dyn_range(uint16_t desired_gyro_dyn_range)
-{
+	return true;
 }
 
 bool
-ADIS16477::self_test()
+ADIS16477::self_test_sensor()
 {
-	DEVICE_DEBUG("self test");
+	DEVICE_DEBUG("self test sensor");
 
 	// self test (global command bit 2)
 	uint8_t value[2] = {};
 	value[0] = (1 << 2);
 	write_reg16(GLOB_CMD, (uint16_t)value[0]);
+	up_mdelay(14); // Self Test Time
```

Comment:
What it the thread of execution here?

Reply:
At the moment it's called from the main that starts the driver, but the idea was too eventually be able to check health periodically and fully reset if necessary during normal cycles (currently an ISR).

Reply:
@dagar 
On the command line a usleep will guarantee to sleep at least the amount (but may be much more), on a ISR usleep can not be used. 

It may be best to design a state machine and not add delays in ISR's (reschedule) - - once we have that pattern can then use it everywhere.